weatherAplication/Modules/Search/ViewModel/SearchViewModel.swift

```diff
@@ -0,0 +1,53 @@
+import Foundation
+import UIKit
+
+class SearchViewModel {
+
+    var delegate: SearchViewModelDelegate?
+    let apiClient = APIClient.sharedInstance
+
+    var APIrequestFlag: Bool = false
+    var cityInfoArray: [CityInfo] = []
+    let delay = 0.5
+
+    func cityChosen(cityInfo: CityInfo) {
+        delegate?.cityChosen(cityInfo: cityInfo)
+    }
+
+    @objc
+    func loadCityNames(sender: UITextField) {
+        if APIrequestFlag != true {
```

Comment:
powinienes unikac wszelkich odwolan do UIKita w ViewModelu, w tym przypadku powinno byc func loadCityNames(), bez @objc. funkcje do #selectora powinienes stworzyc w ViewControllerze, a wywolywalbys w niej ta funkcje fetchujaca miasta + funkcje walidujace, ktorych wynik informowalby o tym czy bedziesz wykonywal requst czy nie (najlepiej zgrupowac oba rodzaje funkcji w jedna funkcje w ViewModelu). w zasadzie powinienes wydzielic metode ktora pobiera miasta unconditionally, a sprawdzanie APIRequestFlag i innych warunkow robic w oddzielnej funkcji/funkcjach, mocno to ulatwi ci potem dodawanie nowych funkcjonalnosci.

Reply:
tutaj dodatkowo jest taki problem, ze powinienes wykonywac request, tak jak zrobiles, po pewnym czasie od zmiany inputu, ale dodatkowo moga byc sytuacje w ktorych w ogole nie powinienes go probowac wykonywac. najwazniejszym przykladem takiej sytuacji jest pusty input - wtedy powinienes po prostu czyscic cityArray